Attach partition...

arthurpassos opened the pull request:
Algorithm is divided into 2 steps:

1. Partition compatibility check;
2. Clone parts and update files;

**Partition compatibility check**
In order for 2 partition expressions to be considered compatible, they must meet one of the following:
1. Destination partition expressions must be a subset of source partition expressions; or
2. Destination partition expression must be monotonically increasing in the source partition min max range.

Monotonicity check does not work for functions with more than 1 variable argument. Therefore, it is a must that all expressions in the partition expression contain at most 1 variable argument.

**Clone**

1. Hard link all files as usual.
2. Delete/ unlink checksums.txt, partition.dat and minmax_<partition_exp_columns>.idx (dst table only).
3. Recalculate those files and write them back in the destination folder.
4. Do rest as usual.

### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow partitions from tables with different partition expressions to be attached when the destination table partition expression doesn't re-partition/ split the part.

Closes #13826 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

arthurpassos committed:
Revert "Revert "Merge pull request #59122 from arthurpassos/bring_back_attach_partition_pr""

This reverts commit 77e4b0c8d909baea75c7701a77f9cb0ee30db3ca.

arthurpassos committed:
use fsync after insert

arthurpassos committed:
add experimental feature setting

arthurpassos committed:
update stateless test

arthurpassos committed:
add test to validate setting is a must

arthurpassos committed:
crazy stuff

arthurpassos committed:
fix replace empty partition

arthurpassos committed:
Revert "crazy stuff"

This reverts commit 5e95dad2c341464c30861a276d9faeed5790706b.

arthurpassos committed:
pre compute partition and partition id

arthurpassos commented:
@tavplubix can you add the can be tested label? No need to review it for now, will improve code further and more tests. Just need a PR with CI running to get a better internal review.

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #60023 from arthurpassos/ClickHouse:fix_59688 -->
*This is an automated comment for commit 36dd7722117d4e01a350776e06f8ffb54583efd8 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60023/36dd7722117d4e01a350776e06f8ffb54583efd8/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60023/36dd7722117d4e01a350776e06f8ffb54583efd8/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

arthurpassos committed:
remove unrelated file

arthurpassos committed:
forbid replace different expression

arthurpassos committed:
add support for move

arthurpassos commented on src/Storages/MergeTree/MergeTreeData.cpp:
```diff
@@ -7023,23 +7012,45 @@ MergeTreeData & MergeTreeData::checkStructureAndGetMergeTreeData(IStorage & sour
         throw Exception(ErrorCodes::BAD_ARGUMENTS, "Tables have different primary key");
 
+
+    if (queryToStringNullable(my_snapshot->getPartitionKeyAST()) != queryToStringNullable(src_snapshot->getPartitionKeyAST()))
```
At this point, there is no distinction if it's attach / replace / move. Maybe it should be improved

arthurpassos committed:
fix

arthurpassos committed:
add tests for move with different partition expressions

arthurpassos commented:
@tavplubix I opted for not supporting replace if partition expressions are different. It doesn't make much sense to support it imho, the partition ids are not related. In which scenario would it be useful and how would the user know which partition would be replaced? I saw you suggested having a new syntax, but even that seems complicated and error prone for the user.

What do you think?

In the other hand, I think it's totally fine to support `move`.

tavplubix commented:
Yes, it makes sense

arthurpassos committed:
add test to validate new parts can be merged

arthurpassos committed:
refactor function, add todo to fix

arthurpassos commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -8103,7 +8191,18 @@ void StorageReplicatedMergeTree::movePartitionToTable(const StoragePtr & dest_ta
+    auto [destination_partition, destination_partition_id] = MergeTreePartitionCompatibilityVerifier::getDestinationPartitionAndPartitionId(
+        is_partition_exp_the_same,
+        src_data,
+        *dest_table_storage, {},
```
I need the data parts to verify partitions are compatible and create destination partition id, but the parts are fetched later using drop range. Drop range depends on destination partition id.

arthurpassos committed:
fix buil

arthurpassos commented on src/Storages/StorageMergeTree.cpp:
```diff
@@ -2050,6 +2050,30 @@ PartitionCommandsResultInfo StorageMergeTree::attachPartition(
     return results;
 }
 
+auto get_destination_partition_and_partition_id =
```
Already changed this, but haven't pushed it yet

arthurpassos commented on src/Storages/MergeTree/MergeTreeData.cpp:
```diff
@@ -4888,7 +4930,15 @@ void MergeTreeData::checkAlterPartitionIsPossible(
+                    // call to `getPartitionIDFromQuery` using source storage.
+                    // Note: `PartitionCommand::REPLACE_PARTITION` is used both for `REPLACE PARTITION` and `ATTACH PARTITION FROM` queries.
+                    // But not for `ATTACH PARTITION` queries.
+                    if (command.type != PartitionCommand::REPLACE_PARTITION)
```
Check `MOVE_PARTITION`

arthurpassos committed:
initial move tests

arthurpassos commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -8114,22 +8206,30 @@ void StorageReplicatedMergeTree::movePartitionToTable(const StoragePtr & dest_ta
-                            covering_part->name, drop_range_fake_part_name);
+                            covering_part->name, source_drop_range_fake_part_name);
+
+        // TODO fix below
```
Ideally, this would be done only once (as it is being done in other methods `replacePartitionFrom` and `movePartitionToTable` on non-replicated engine. The problem is that in order to verify compatibility between partitions and calculate new partition id, the source parts are required. Source parts are grabbed only inside the loop and I am still not sure if this logic can be moved out of here.

arthurpassos commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -8235,9 +8360,9 @@ void StorageReplicatedMergeTree::movePartitionToTable(const StoragePtr & dest_ta
 
         /// Cancel concurrent inserts in range
-        clearLockedBlockNumbersInPartition(*zookeeper, drop_range.partition_id, drop_range.min_block, drop_range.max_block);
+        clearLockedBlockNumbersInPartition(*zookeeper, source_drop_range.partition_id, source_drop_range.min_block, source_drop_range.max_block);
```
This requires extra attention. I am using the source drop range, which means the source partition id is in the drop range.

The reasoning behind this decision was: `move partition` does not delete anything on destination, so all drops/deletes will happen on source. Which means source id is the one to be used.

arthurpassos commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -8207,14 +8332,14 @@ void StorageReplicatedMergeTree::movePartitionToTable(const StoragePtr & dest_ta
         ReplicatedMergeTreeLogEntryData entry;
         {
-            MergeTreePartInfo drop_range_dest = makeDummyDropRangeForMovePartitionOrAttachPartitionFrom(partition_id);
+            MergeTreePartInfo drop_range_dest = makeDummyDropRangeForMovePartitionOrAttachPartitionFrom(destination_partition_id);
```
Destination partition id instead of source

arthurpassos commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -8329,7 +8454,7 @@ void StorageReplicatedMergeTree::movePartitionToTable(const StoragePtr & dest_ta
 
         /// Cleaning possibly stored information about parts from /quorum/last_part node in ZooKeeper.
-        cleanLastPartNode(partition_id);
+        cleanLastPartNode(destination_partition_id);
```
This might be wrong

arthurpassos committed:
workaroun

arthurpassos committed:
tuple to nonpartitioned0 and the other way around

arthurpassos committed:
test ci

arthurpassos committed:
..

arthurpassos committed:
remove replicated tests to see if fuzzer passes

arthurpassos committed:
attempt to fix replicated move partition

arthurpassos committed:
add a few more move test

arthurpassos committed:
integ tests for move distinct partition exp

arthurpassos committed:
remove some small functions and inline them as previously requested

arthurpassos committed:
fix black

arthurpassos committed:
Merge branch 'master' into fix_59688

arthurpassos committed:
Merge branch 'ClickHouse:master' into fix_59688

arthurpassos committed:
hehe

arthurpassos committed:
Merge branch 'master' into fix_59688

arthurpassos committed:
remove dupl incl

arthurpassos commented on src/Storages/StorageMergeTree.cpp:
```diff
@@ -2079,47 +2082,96 @@ void StorageMergeTree::replacePartitionFrom(const StoragePtr & source_table, con
     if (replace)
     {
-        drop_range.partition_id = partition_id;
+        drop_range.partition_id = destination_partition_id;
```
This does not make a difference. It won't reach this code if `replace == true` and `is_partition_expression_the_same == false`

arthurpassos commented on src/Storages/MergeTree/MergeTreePartitionCompatibilityVerifier.cpp:
```diff
@@ -0,0 +1,135 @@
+    return partition;
+}
+
+std::pair<MergeTreePartition, std::string> MergeTreePartitionCompatibilityVerifier::getDestinationPartitionAndPartitionId(
```
I can also return `std::pair<std::optional<MergeTreePartition>, std::string>`

arthurpassos committed:
improve code a bit

arthurpassos committed:
update docs

arthurpassos committed:
rename method

arthurpassos commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -8128,22 +8218,33 @@ void StorageReplicatedMergeTree::movePartitionToTable(const StoragePtr & dest_ta
+        }
+
+        // Can I safely return here? This is done in all other places.
+        if (src_all_parts.empty())
```
I assume it is safe to abort here

arthurpassos committed:
add better error messages for monotonicity check and test that validates monotonicity on specific range

arthurpassos committed:
Revert "improve code a bit"

This reverts commit f52d705e78c1a3ee06a6fff83f05b53edfa64a7f.

arthurpassos commented on src/Interpreters/MonotonicityCheckVisitor.h:
```diff
@@ -33,7 +37,23 @@ class MonotonicityCheckMatcher
-        void reject() { monotonicity.is_monotonic = false; }
+        Range range = Range::createWholeUniverse();
+
+        enum class ErrorCode
```
I can instead throw exceptions. This visitor is used only in two places: `TreeOptimizer` and `MergeTreePartitionCompatibilityVerifier`. I can add a try catch on former if necessary.

arthurpassos committed:
trigger ci

arthurpassos committed:
fix a few column mismatch issues

arthurpassos committed:
rely on ci

arthurpassos committed:
update tests

arthurpassos committed:
rename a few things

arthurpassos committed:
use 3 nodes and clustering for integ tests

arthurpassos committed:
fix black

arthurpassos committed:
Merge branch 'ClickHouse:master' into fix_59688

arthurpassos commented:
@tavplubix I have fixed all the issues you mentioned and a few more I found on the way.

Summary:

1. Partition id mess should be fixed
2. Move partition support added
3. Removed functions that abstracted a one liner or very few lines in `MergeTreeDataPartCloner`
4. Added more tests, specially move partition
5. Added support for `fsync`
6. Added experimental feature setting
7. Exception for replace partition as discussed
8. Better exception messages for monotonicity

The only CI failure is related to flaky checks, but I could not reproduce locally or find the root cause. I have tried the below:

1. Run the command CI runs (without some previous setup the check does) - `./clickhouse-test --binary /home/arthur/ClickHouse/cmake-build-debug-asan/programs/clickhouse --testname --shard --zookeeper --check-zookeeper-session --hung-check --print-time --test-runs 100 --hung-check --print-time 02456_test_move_partition_different_partition_exp_replicated. 02456_test_attach_partition_different_partition_exp. 02456_test_move_partition_different_partition_exp. 02456_test_attach_partition_different_partition_exp_replicated. --order=random --jobs 8 --report-logs-stats`. It succeeds.
9. Set the same settings manually (both user session and merge tree) - tests simply fail because it does not reflect .reference file. Running these queries manually did not show any significant difference.

Do you have any suggestions on how to debug this?

Also, can you re-review it?

arthurpassos committed:
fix conflicts

arthurpassos committed:
fix conflicts pt2

arthurpassos commented:
@tavplubix can you take a look?

arthurpassos committed:
Merge branch 'master' into fix_59688

arthurpassos committed:
integrate attach from other disk

arthurpassos committed:
fix style

arthurpassos committed:
normalize err messages & fix clang tidy

arthurpassos commented:
@tavplubix kind reminder

arthurpassos committed:
fix dupl include

arthurpassos committed:
remove unnecessary sync

arthurpassos committed:
remove from ref file as well

arthurpassos committed:
use long

arthurpassos commented:
@alexey-milovidov I suppose Tokmakov is busy and/or has other priorities. Maybe you or someone you know can take a look at this one?

alexey-milovidov commented:
We have a few tests that failed after enabling Analyzer - as soon as they are fixed, we can continue.

arthurpassos commented:
> We have a few tests that failed after enabling Analyzer - as soon as they are fixed, we can continue.

Hi @alexey-milovidov. Any estimates on when that'll be completed or a place where progress can be tracked?

arthurpassos committed:
Merge branch 'master' into fix_59688

alexey-milovidov commented:
Here: https://aretestsgreenyet.com/

alexey-milovidov commented:
We have noticed a failure of this test in master: https://s3.amazonaws.com/clickhouse-test-reports/0/30725bcd7a9eab3332ef15b2f6d84e9226d0d38e/integration_tests__tsan__[2_6].html

alexey-milovidov commented:
PTAL https://github.com/ClickHouse/ClickHouse/pull/53286#issuecomment-2251416434
